receiver/awsfirehosereceiver/internal/unmarshaler/cwmetricstream/metricsbuilder.go

```diff
@@ -0,0 +1,173 @@
+	attributes.InsertString(conventions.AttributeCloudAccountID, rmb.accountID)
+	attributes.InsertString(conventions.AttributeCloudRegion, rmb.region)
+	splitNamespace := strings.SplitN(rmb.namespace, namespaceDelimiter, 2)
+	if len(splitNamespace) == 2 && strings.EqualFold(splitNamespace[0], conventions.AttributeCloudProviderAWS) {
+		attributes.InsertString(conventions.AttributeServiceNamespace, splitNamespace[0])
+		attributes.InsertString(conventions.AttributeServiceName, splitNamespace[1])
+	} else {
+		attributes.InsertString(conventions.AttributeServiceName, rmb.namespace)
+	}
+	attributes.InsertString(attributeAWSCloudWatchMetricStreamName, rmb.metricStreamName)
+}
+
+// The metricBuilder aggregates metrics of the same name and unit
+// into data points. Stores the timestamps for each added metric
+// in a set to prevent duplicates.
+type metricBuilder struct {
+	// name is the metric name.
+	name string
+	// unit is the metric unit.
+	unit string
+	// dataPoints is the slice of summary data points
+	// for the metric.
+	dataPoints pdata.SummaryDataPointSlice
+	// timestamps is the set of seen timestamps.
+	timestamps map[string]bool
+}
+
+// newMetricBuilder creates a metricBuilder with the name and unit.
+func newMetricBuilder(name, unit string) *metricBuilder {
+	return &metricBuilder{
+		name:       name,
+		unit:       unit,
+		dataPoints: pdata.NewSummaryDataPointSlice(),
+		timestamps: make(map[string]bool),
+	}
+}
+
+// AddDataPoint adds the metric as a datapoint if a metric for that timestamp
+// hasn't already been added.
+func (mb *metricBuilder) AddDataPoint(metric cWMetric) {
+	key := mb.toTimestampKey(metric)
+	if _, ok := mb.timestamps[key]; !ok {
+		mb.toDataPoint(mb.dataPoints.AppendEmpty(), metric)
+		mb.timestamps[key] = true
+	}
+}
+
+// Build builds the pdata.Metric with the data points that were added
+// with AddDataPoint.
+func (mb *metricBuilder) Build(metric pdata.Metric) {
+	metric.SetName(mb.name)
+	metric.SetUnit(mb.unit)
+	metric.SetDataType(pdata.MetricDataTypeSummary)
+	mb.dataPoints.MoveAndAppendTo(metric.Summary().DataPoints())
+}
+
+// toDataPoint converts a cWMetric into a pdata datapoint and attaches the
+// dimensions as attributes.
+func (mb *metricBuilder) toDataPoint(dp pdata.SummaryDataPoint, metric cWMetric) {
+	dp.SetCount(uint64(metric.Value.Count))
+	dp.SetSum(metric.Value.Sum)
+	qv := dp.QuantileValues()
+	min := qv.AppendEmpty()
+	min.SetQuantile(0)
+	min.SetValue(metric.Value.Min)
+	max := qv.AppendEmpty()
+	max.SetQuantile(1)
+	max.SetValue(metric.Value.Max)
+	dp.SetTimestamp(pdata.NewTimestampFromTime(time.UnixMilli(metric.Timestamp)))
+	for k, v := range metric.Dimensions {
+		dp.Attributes().InsertString(ToSemConvAttributeKey(k), v)
+	}
+}
+
+// toTimestampKey combines the dimensions and timestamps to create a key
+// used to prevent duplicate metrics.
+func (mb *metricBuilder) toTimestampKey(metric cWMetric) string {
+	return fmt.Sprintf("%v::%v", metric.Dimensions, metric.Timestamp)
```

Comment:
Better to use a struct than a string for deduping, also we might need to sort the dimensions (other cloudwatch components use a `sorted.Label` lib from the otel sdk to handle that within a struct I think)

Reply:
Added a `dataPointKey` struct with the timestamp and string dimensions. From [Go 1.12](https://go.dev/doc/go1.12#minor_library_changes), the fmt library prints dimensions in key-sorted order. Considered using `attributes.Distinct` instead, but involved a lot more allocation than seemed necessary and could not be reused for the `pdata.AttributeMap`.

```go
        var dimensions []attribute.KeyValue
	for k, v := range metric.Dimensions {
		dimensions = append(dimensions, attribute.String(k, v))
	}
	set := attribute.NewSetWithSortable(dimensions, sortable)
	return dataPointKey{
		timestamp: metric.Timestamp,
		dimensions: set.Equivalent(),
	}
```

Reply:
Thanks - I guess someday we'd want a pdata version of that